### ISSUE TITLE
imb

### ISSUE THREAD

**Issue by gjse** [gjse/project#6]
#include "mbed.h"
#include "pw_mode.h"
#include "Adafruit_SSD1306.h"
#include "melody.h"

extern Serial pc;

I2C i2c1(I2C_SDA, I2C_SCL);
Adafruit_SSD1306_I2c myOled(i2c1, D13, 0x78, 64, 128);

static int storedPW[4] = { 1, 0, 0, 0};

static int pw[4] = { 0, 0, 0, 0};
static int ppw[4] = { -1, -1, -1, -1};

static int xp[4] = { 1, 32, 64, 96};
static int yp = 21;

void displayPassword() {
	bool bRefreshed = false;
	
	for(int i=0; i<4; i++){
		if(pw[i] != ppw[i]){
			myOled.setTextCursor(xp[i],yp);
			myOled.writeChar(0x30 + pw[i]);
			ppw[i] = pw[i];
			bRefreshed = true;
			beep();
		}
	}
	
	if (bRefreshed){
		myOled.display();
	}
}

static int currCursorPos = 0;
static int prevCursorPos = -1;

#define CURSOR_Y_POS 	61
#define CURSOR_WIDTH 	26
#define CURSOR_HEIGHT	2

void displayCursor() {
	if (currCursorPos != prevCursorPos){
		if(prevCursorPos >= 0){
			myOled.fillRect(xp[prevCursorPos]+1,
			CURSOR_Y_POS, CURSOR_WIDTH, CURSOR_HEIGHT, 0);
		}
		
		myOled.fillRect(xp[currCursorPos]+1,
			CURSOR_Y_POS, CURSOR_WIDTH, CURSOR_HEIGHT, 1);
		myOled.display();
		
		prevCursorPos = currCursorPos;
	}
}

static AnalogIn xAxis(PC_2);
static AnalogIn yAxis(PC_3);

typedef	enum {NEUTRAL=0, LEFT=1, RIGHT=-1, UP=1, DOWN=-1} js_value_t;
static volatile js_value_t xv, yv;

#define CENTERAL_VALUE	69

void getJoystickValue(){
	int x = int(xAxis.read() * 100);
	int y = int(yAxis.read() * 100);
	//pc.printf("x : %d, y : %d \t", x, y);
	x = CENTERAL_VALUE - x;
	if(abs(x) < 4) xv = NEUTRAL;
	else if (x > 0) xv = LEFT;
	else xv = RIGHT;
	
	y = CENTERAL_VALUE - y;
	if(abs(y) < 4) yv = NEUTRAL;
	else if (y > 0) yv = UP;
	else yv = DOWN;
	//pc.printf("X : %d, Y : %d \t", x, y);
	
}

void changePWandCursor() {
	if (yv != NEUTRAL){
		int num = pw[currCursorPos];
		if(yv == UP){
			if (++num > 9) num = 0;
		} 
		else if (yv == DOWN){
			if (--num < 0) num = 9;
		}
		pw[currCursorPos] = num;
	} 
	else if (xv != NEUTRAL){
		prevCursorPos = currCursorPos;
		if (xv == LEFT){
			if(--currCursorPos < 0) currCursorPos = 3;
		}
		else if (xv == RIGHT){
			if(++currCursorPos > 3) currCursorPos = 0;
		}
	}
	//pc.printf("X : %s, Y : %s \n", (xv==1)? "LEFT":"RIGHT", (yv==1)?"UP":"DOWN");
}

static Ticker jsTicker;
static Ticker pwTicker;

static DigitalIn enterBtn(PB_7);
static DigitalOut doorLed(PA_13);

bool isEnterBtnPressed(){
	static int prevState = 1;
	int currState = enterBtn.read();
	
	if (currState != prevState){
		wait_ms(30); // debouncing delay
		prevState = currState;
		if (currState == 0){
			return true;
		}
	}
	return false;
}

bool checkPassword(){
	bool matched = true;
	
	for(int i=0;i<4;i++){
		if(storedPW[i] != pw[i]){
			matched = false;
			//wrongMelody();
			break;
		}
	}
		
	return matched;
}

void initPWMode(){
	// initialize OLED display
	myOled.clearDisplay();
	
	myOled.setTextSize(1);
	myOled.setTextCursor(0, 0);
	myOled.printf("[ PW Mode ] \r\n");
	myOled.display();
	
	
	myOled.setTextSize(5);
	
	for(int i=0; i<4; i++){
		pw[i] = 0;
		ppw[i] = -1;		
	}
	currCursorPos = 0;
	prevCursorPos = -1;
	
	jsTicker.attach(&getJoystickValue, 0.2);
	pwTicker.attach(&changePWandCursor, 0.4);
}

void cleanupPWMode(){
	jsTicker.detach();
	pwTicker.detach();
}

void doPWModeOperation(){
	displayPassword();
	displayCursor();
	
	if(isEnterBtnPressed()){
		if(checkPassword()){
			// open door
			doorLed = !doorLed;
		}
	}
}
